packages/compadre/CMakeLists.txt

```diff
@@ -0,0 +1,617 @@
+CMAKE_MINIMUM_REQUIRED(VERSION 3.10.0 FATAL_ERROR)
+
+set(Compadre_VERSION "1.0.3")
+
+# Called at the top of every CMakeLists.txt file
+macro(include_tribits_build)
+  if (${PROJECT_NAME}_TRIBITS_DIR)
+    include("${CMAKE_CURRENT_SOURCE_DIR}/CMakeLists.tribits.cmake")
+    return()
+  endif()
+endmacro()
+
+# cmake helper functions
+include(${CMAKE_CURRENT_LIST_DIR}/cmake/bob.cmake)
+
+if (${PROJECT_NAME}_TRIBITS_DIR) # TriBITS CMake Project
+
+  #
+  # A) Define the package
+  #
+
+  tribits_package(Compadre)
+
+  #
+  # B) Set up package-specific options
+  #
+  TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_DEBUG
+      Compadre_DEBUG
+      "Enable debug code in compadre"
+      ${Trilinos_ENABLE_DEBUG} )
+
+  TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_EXTREME_DEBUG
+      Compadre_EXTREME_DEBUG
+      "Enable extreme debug code in compadre"
+      OFF )
+
+  TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_LAPACK_DECLARED_THREADSAFE
+      LAPACK_DECLARED_THREADSAFE
+      "Set to OFF if non-threadsafe version of LAPACK being used"
+      ON )
+
+  TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_USE_LAPACK
+      Compadre_USE_LAPACK
+      "Either Compadre_USE_LAPACK, Compadre_USE_MKL, or Compadre_USE_CUDA must be set to ON. LAPACK is default."
+      OFF )
+
+  TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_USE_MKL
+      Compadre_USE_MKL
+      "Either Compadre_USE_LAPACK, Compadre_USE_MKL, or Compadre_USE_CUDA must be set to ON. LAPACK is default."
+      OFF )
+
+  TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_USE_CUDA
+      Compadre_USE_CUDA
+      "Either Compadre_USE_LAPACK, Compadre_USE_MKL, or Compadre_USE_CUDA must be set to ON. LAPACK is default."
+      OFF )
+
+  # ensure only one of Compadre_USE_LAPACK, Compadre_USE_MKL, or Compadre_USE_CUDA are ON
+  if (Compadre_USE_LAPACK)
+    set(Compadre_USE_MKL OFF)
+    set(Compadre_USE_CUDA OFF)
+  elseif (Compadre_USE_MKL)
+    set(Compadre_USE_CUDA OFF)
+  endif() #if Compadre_USE_CUDA, all others already set to OFF
+  # if all off (default), then set Compadre_USE_LAPACK to ON
+  if ((NOT(Compadre_USE_LAPACK) AND NOT(Compadre_USE_MKL)) AND NOT(Compadre_USE_CUDA))
+    set(Compadre_USE_LAPACK ON)
+  endif()
+
+  # ensure TPL_ENABLE_*** configured consistent with Compadre_USE_***
+  if(Compadre_USE_LAPACK AND NOT(TPL_ENABLE_LAPACK))

```

Comment:
I would argue that these should trigger a `FATAL ERROR` instead of a message. It is faster to explicitly disable compadre at configure time to fix this issue than to wait for a linking error after Trilinos has been built.

Reply:
@lucbv, I only disable the package if the user doesn't enable the TPL. If I add a FATAL_ERROR, cmake will fail. Is this the desired behavior? I thought if a packages TPLs can't be satisfied, then the package should just be turned off?

Reply:
I do not know if Trilinos has a set rule for that?
I just feel that if you explicitly enable a package at configure time and it then gets disabled you might end up compiling Trilinos (sometime a slow process...) and only realize that there is an issue if you link against an executable.
I would rather bug the user and have him/her fix the configuration line right away.

Reply:
@bartlettroscoe can you please weigh in on this? Thanks!

Reply:
The Dependencies.cmake requirements already are conditional on Compadre_USE_LAPACK, Compadre_USE_MKL, or Compadre_USE_CUDA. Each sets a different combination of required TPLs (the ones being checked against here). Theoretically, it isn't possible for the user to hit this case. This is just a double check that required TPLs are present.

Reply:
I'm still interested to know what is the desired behavior for a Trilinos package, but I think there is no circumstance under which this could get called.

Reply:
Oh right, if cmake has a required dependency on LAPACK then I think we are good. @lucbv are you fine with this?

Reply:
> @bartlettroscoe can you please weigh in on this? Thanks!

Looking at the Dependencies.cmake file, I am not sure this this is consistent with how the rest of Trilinos deals with BLAS and LAPACK (and MKL).

Looking at:

```
$ find packages/ -name Dependencies.cmake -exec grep -l MKL {} \;
packages/amesos/cmake/Dependencies.cmake
packages/amesos2/cmake/Dependencies.cmake
packages/muelu/cmake/Dependencies.cmake
packages/kokkos-kernels/cmake/Dependencies.cmake
packages/shylu/shylu_node/basker/cmake/Dependencies.cmake
packages/shylu/shylu_node/tacho/cmake/Dependencies.cmake
packages/shylu/shylu_node/hts/cmake/Dependencies.cmake
packages/shylu/shylu_dd/bddc/cmake/Dependencies.cmake
packages/stokhos/cmake/Dependencies.cmake
```

it does not look like MKL is being used as a replacement for BLAS and LAPACK (and it can't if you think about it).

I did not have anything to do with the definition of the 'MKL' TPL or its usage in any package.  I just know that I have never seen it enabled in any of the projects I have worked with.

Looking at a few examples, it looks like it is mostly used for testing and not as an implementation for BLAS.

I would just stick with the BLAS and LAPACK TPLs and only define a dependency on MKL if you are using functions in MKL that go beyond BLAS and LAPACK.

Note that there is no such thing as "FindMKL.cmake" as shown at:

* https://cmake.org/cmake/help/v3.17/manual/cmake-modules.7.html

Looks like someone attempted to contribute a MKLConfig.cmake file to Intel:

* https://software.intel.com/en-us/forums/intel-math-kernel-library/topic/850824

but I don't think that went anywhere.

Reply:
Can the MKL be left in this PR as an option (it has to be explicitly set to ON, and it is not currently used [future use])?

@bartlettroscoe if a required TPL for a package wasn't found, would it be better to throw a fatal error or to turn off the package?

Reply:
> @bartlettroscoe if a required TPL for a package wasn't found, would it be better to throw a fatal error or to turn off the package?

@kuberry, if a required TPL can't be found, it is a fatal configure error.

My advice is to just have the simpler Dependencies.cmake file:

```
if (TPL_ENABLE_CUDA)
  set(EXTRA_REQUIRED_TPLS CUSOLVER CUBLAS)
endif()

tribits_package_define_dependencies(
  LIB_REQUIRED_PACKAGES
    KokkosCore KokkosContainers KokkosAlgorithms
  LIB_REQUIRED_TPLS
    LAPACK BLAS ${EXTRA_REQUIRED_TPLS}
  LIB_OPTIONAL_TPLS
    MPI CUDA
  )
```

Just leave it to TriBITS and the user configuring Trilinos on providing BLAS and LAPACK through MKL.

Don't bother with the MKL TPL unless you are using functionality in MKL beyond just what is in BLAS and LAPACK.

It is possible that way that the file compadre/cmake/Dependencies.cmake is currently defined, it might actually be broken when BLAS, LAPACK, and MKL are all enabled.  Has anyone actually tested this with the BLAS, LAPACK,and MKL enabled?

Reply:
@bartlettroscoe I removed all MKL logic since the package doesn't currently use it. This package either uses LAPACK+BLAS or CUBLAS+CUSOLVER. It doesn't use LAPACK routines if using CUSOLVER. I think what I have in Dependencies.cmake now should take care of this.

@lucbv I removed the double check for if TPLs were enabled since they must be (they are required TPLs in Dependencies.cmake). Effectively, since the TPLs are listed as required, if they are not enabled/found then the configure will fail (what you mentioned as the desired behavior).

I think this addresses everyone's comments/input. Thank you both for the input.